shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java

```diff
@@ -0,0 +1,342 @@
+    private ResourceExtractor resourceExtractor;
+    @Nullable
+    private Settings settings;
+
+    /**
+     * Starts initialization of the native system.
+     * @param applicationContext The Android application context.
+     */
+    public void startInitialization(@NonNull Context applicationContext) {
+        startInitialization(applicationContext, new Settings());
+    }
+
+    /**
+     * Starts initialization of the native system.
+     * <p>
+     * This loads the Flutter engine's native library to enable subsequent JNI calls. This also
+     * starts locating and unpacking Dart resources packaged in the app's APK.
+     * <p>
+     * Calling this method multiple times has no effect.
+     *
+     * @param applicationContext The Android application context.
+     * @param settings Configuration settings.
+     */
+    public void startInitialization(@NonNull Context applicationContext, @NonNull Settings settings) {
+        if (Looper.myLooper() != Looper.getMainLooper()) {
+          throw new IllegalStateException("startInitialization must be called on the main thread");
+        }
+        // Do not run startInitialization more than once.
+        if (this.settings != null) {
+          return;
+        }
+
+        this.settings = settings;
+
+        long initStartTimestampMillis = SystemClock.uptimeMillis();
+        initConfig(applicationContext);
+        initResources(applicationContext);
+
+        System.loadLibrary("flutter");
+
+        VsyncWaiter
+            .getInstance((WindowManager) applicationContext.getSystemService(Context.WINDOW_SERVICE))
+            .init();
+
+        // We record the initialization time using SystemClock because at the start of the
+        // initialization we have not yet loaded the native library to call into dart_tools_api.h.
+        // To get Timeline timestamp of the start of initialization we simply subtract the delta
+        // from the Timeline timestamp at the current moment (the assumption is that the overhead
+        // of the JNI call is negligible).
+        long initTimeMillis = SystemClock.uptimeMillis() - initStartTimestampMillis;
+        FlutterJNI.nativeRecordStartTimestamp(initTimeMillis);
+    }
+
+    /**
+     * Blocks until initialization of the native system has completed.
+     * <p>
+     * Calling this method multiple times has no effect.
+     *
+     * @param applicationContext The Android application context.
+     * @param args Flags sent to the Flutter runtime.
+     */
+    public void ensureInitializationComplete(@NonNull Context applicationContext, @Nullable String[] args) {
+        if (Looper.myLooper() != Looper.getMainLooper()) {
+          throw new IllegalStateException("ensureInitializationComplete must be called on the main thread");
+        }
+        if (settings == null) {
+          throw new IllegalStateException("ensureInitializationComplete must be called after startInitialization");
+        }
+        if (initialized) {
```

Comment:
optional: Same here about floating the return about the exceptions.

Reply:
Done.